web/src/hashchange.js

```diff
@@ -201,6 +201,13 @@ function do_hashchange_normal(from_reload) {
                     narrow_opts.then_select_offset = page_params.initial_narrow_offset;
                 }
             }
+            const location_data_for_hash =
+                browser_history.hash_to_page_location[window.location.hash];
+            if (location_data_for_hash !== undefined) {
+                message_lists.home.pre_narrow_offset = location_data_for_hash.offset;
+                narrow_opts.then_select_id = location_data_for_hash.narrow_pointer;
+                narrow_opts.then_select_offset = location_data_for_hash.narrow_offset;
+            }
```

Comment:
So what I had in mind was not this, which only would help for the specific case of clicking on a message in search -- but instead making it the case that whenever we __leave__ any narrow through any mechanism, we save the state such that hitting the `back` button, potentially repeatedly, takes you to the offset within that narrow where you were. 

It might still be reasonable to merge this class of change just to get some experience with this type of approach, but we should at least put a `TODO` comment here explaining that context

Reply:
Oh I see! Yes I think that would be a lot better, for several reasons.

I'll see how easy it is to do that, and if there are complications then maybe we can merge this as an intermediate fix for one of the more annoying cases.

Reply:
Started a conversation about this here: https://chat.zulip.org/#narrow/stream/6-frontend/topic/browser.20navigation/near/1621157